File: pe/visualstudio/PE.CSharp/Problem0004.cs
using System.Threading.Tasks;

namespace PE.CSharp
{
    public class Problem0004
    {
        public bool IsPalendrom(int n)
        {
            var nString = n.ToString();
            
            var stringLength = nString.Length;
            var index = 0;
            
            var isPalendrom = true;

            while (isPalendrom && (stringLength - (index * 2) > 1))
            {
                if (nString[index] != nString[stringLength - (index + 1)])
                {
                    isPalendrom = false;
                }

                index++;
            }

            return isPalendrom;
        }

        public int MaxPalendromSync(int start, int end)
        {
            var maxPalendrom = 0;
            
            for (var i = start; i < end + 1; i++)
            {
                for (var j = i; j < end + 1; j++)
                {
                    var product = j*i;
                    if (product > maxPalendrom && IsPalendrom(product))
                    {
                        maxPalendrom = product;
                    }
                }
            }

            return maxPalendrom;
        }

        public int MaxPalendromParallel(int start, int end)
        {
            var maxPalendrom = 0;

            Parallel.For(start, end + 1, i => Parallel.For(i, end + 1, j =>
            {
                var product = j*i;
                if (product > maxPalendrom && IsPalendrom(product))
                {
                    maxPalendrom = product;
                }

                //this isn't threadsafe or better than the old solution. 
            }));

            return maxPalendrom;
        }
    }
}

File: pe/visualstudio/PE.JavaScript/Implementation/Problem0004.js
var javascriptPE;
(function (javascriptPE) {
    (function (implementation) {

        var problem0004 = (function() {

            function problem0004() {
                
            }

            problem0004.prototype.isPalendrom = function(n) {
                var nString = n.toString();

                var stringLength = nString.length;
                var index = 0;

                var isPalendrom = true;

                while (isPalendrom && (stringLength - (index * 2) > 1)) {
                    if (nString[index] != nString[stringLength - (index + 1)]) {
                        isPalendrom = false;
                    }

                    index++;
                }

                return isPalendrom;
            };

            problem0004.prototype.maxPalendromProduct = function(start, end) {
                var maxPalendromProduct = 0;

                for (var i = start; i < end + 1; i++) {
                    for (var j = i; j < end + 1; j++) {
                        var product = j * i;
                        if (product > maxPalendromProduct && this.isPalendrom(product)) {
                            maxPalendromProduct = product;
                        }
                    }
                }

                return maxPalendromProduct;
            };
            


            return problem0004;
        })();
        implementation.problem0004 = problem0004;

    })(javascriptPE.implementation || (javascriptPE.implementation = {}));
})(javascriptPE || (javascriptPE = {}));